---
title: "Statistical analysis"
output:
  pdf_document: default
  html_document: default
date: "2023-03-29"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(car)
library(tidyverse)
library(magrittr)
library(tidymodels)
library(cowplot)
library(modelr)
library(naniar)
library(ggeffects)
library(patchwork)
library(jtools)
```
## Purpose

The goal of this section is to build a model of the number of 16S gene copies

## Reading in the data

```{r}
D <- readRDS("./data/pred_bacdive_growth_ribdif.rds")
D %>% select(last_col(41):last_col()) 
```

## Large analysis
### Data transformations

```{r}
# for Pathogen column, set nan to 0
Dt_tmp1 <- D %>% 
  mutate(pathogen_any = ifelse(is.na(pathogen_any), "0", pathogen_any))

# Set nan to neutrophile for PH.range
Dt_tmp2 <- Dt_tmp1 %>% 
  mutate(PH.range = ifelse(is.na(PH.range), "neutrophile", PH.range))

# Get pseudogenes as a percent
Dt_tmp3 <- Dt_tmp2 %>% 
  mutate(pseudogenes_percent = 100*(pseudogenes/total_genes))

# calculate surface/area + set motility which is NAN to notknown
Dt_tmp3.1 <- Dt_tmp3 %>% 
  mutate(motility = ifelse(is.na(motility), "Not Known", motility)) %>% 
  mutate(surface = 2*(cell.width^2) + 4*(cell.width*cell.length),
         area = cell.width*cell.width*cell.length,
         surf_area = surface/area) %>% 
          select(-surface, -area)


# Set Sample type as the sample from which is is sampled the most often from
Dt_tmp3.2 <- Dt_tmp3.1 %>% 
  rowwise() %>% 
  mutate(
    max = max(pick(ends_with("counts")))
    ) %>% 
  ungroup() %>% 
  mutate(environment = case_when(
      aquatic.counts == max ~ "aquatic.counts",
      soil.counts == max ~ "soil.counts",
      plant.counts == max ~ "plant.counts",
      animal.counts == max ~ "animal.counts")) %>% 
  select(-max)

# Add surface/volume for missing values based on median per family
Dt_tmp3.3 <- Dt_tmp3.2 %>% 
  # Calculate the median
  group_by(family) %>% 
  mutate(median_surf_area = median(surf_area, na.rm=T)) %>% 
  ungroup() %>% 
  # Add the median for the missng values
  mutate(surf_area = ifelse(is.na(surf_area), median_surf_area, surf_area))

# Format oxygen tolerance
Dt_tmp3.4 <- Dt_tmp3.3 %>% 
  mutate(oxygen.tolerance = case_when(
    oxygen.tolerance == "obligate anaerobe" ~ "anaerobe",
    oxygen.tolerance == "obligate aerobe" ~ "aerobe",
    oxygen.tolerance == "facultative anaerobe" ~ "aerobe",
    oxygen.tolerance == "facultative aerobe" ~ "aerobe",
    oxygen.tolerance == "microaerophile" ~ "microaerophile",
    .default = oxygen.tolerance
  ))

# Select the data relevant for the first analysis
# It throws error due to NANS
Dt_tmp4 <- Dt_tmp3.4 %>% 
  select(pathogen_any, PH.range, oxygen.tolerance, 
         growth_temp,  gc_percent, sporeforming,
         ar_count, surf_area, motility, environment,
         n16, phylum, ar_count, genome_components,
         pseudogenes_percent, genes_coding) %>% 
  mutate(growth_temp = as.double(growth_temp)) 

```

```{r}


# Removing the missing data for modelling
Dt_tmp5 <- Dt_tmp4 %>% 
  filter(if_all(everything(), ~!is.na(.x)))

# Removing phylums with < 10 entries
Dt_tmp6 <- Dt_tmp5 %>% 
  mutate(n=n(), .by=phylum) %>% 
  filter(n > 10) %>% 
  select(-n)

```


# Transformations

```{r}

# Apply the transformations
Dt_tmp7 <- Dt_tmp6  %>% 
  mutate(across(c(n16, genome_components, pseudogenes_percent, genes_coding), log2))

# And filter out phylums with less than 10 entries
Dt <- Dt_tmp7 %>% 
  mutate(n=n(), .by=phylum) %>% 
  filter(n>10) %>% 
  select(-n)

```

Make  function to backtransform data

```{r}
# Function to back transform data
back_transform <- function(df){
  df %>% 
    mutate(across(c(n16, genome_components, pseudogenes_percent, genes_coding), ~2^.x))
}

```


# Make the model
Fit model
```{r}
fit <- lm(n16 ~ ., data=Dt)
```

Step transform it 
```{r}

after_step <- stats::step(fit,  scope=~.^4, direction = "both", test="F", k=log(nrow(Dt)))
Anova(after_step)
summ(after_step)
fit_fn <- after_step

```


Looking at diastognistic plot
```{r}
par(mfrow=c(2,2))
plot(fit_fn)
```

On the top left plot, each line represents a number of 16s gene copies. It seems to always predict wrong on n16=1 and often on n16=2
Finding the one with the high leverage
In the above plots we see 681,305,865 have sqrt res over 3. Lets inspect them
```{r}
Dt[c(681,305,865),] %>% back_transform()
# Could have sf pred wrong?
```
They seem to vary a lot. There is no clear pattern.
Although the values seem ok, they have to much to say compared to other points


## Model without outliers

```{r}
# Remove the outliers
Dt_no_outliers <- Dt %>% filter(!row_number() %in% c(681,305,865)) %>% 
  mutate(sporeforming = as.factor(sporeforming))

# fit model again
fit <- lm(n16 ~ ., data=Dt_no_outliers)

```

Step transform it 
```{r}
after_step <- stats::step(fit,  scope=~.^4, direction = "both", test="F", k=log(nrow(Dt)))
anova(after_step)
summary(after_step)
fit_fn <- after_step

```

ANCOVA of model
```{r}
Anova(fit_fn) 
```

--
Looking at diagnostic plot
```{r}
par(mfrow=c(2,2))
plot(fit_fn)
```
Seems a lot better. This is the final model



Visualizing it
```{r}

add_predictions(Dt, fit_fn) %>% 
  ggplot(aes(x=2^genes_coding, y=2^pred, col="red")) +
  geom_point() +
  geom_point(aes(y=2^n16), col="grey") +
  facet_wrap(~phylum)

```
predictions look good
