---
title: "Effect plots"
output:
  pdf_document: default
  html_document: default
date: "2023-03-29"
---


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(car)
library(tidyverse)
library(magrittr)
library(tidymodels)
library(cowplot)
library(modelr)
library(naniar)
library(ggeffects)
library(patchwork)
library(jtools)
library(gtsummary)
library(emmeans)
```
## Purpose

The goal of this section is to plot the effects of the model of the number of 16S gene copies. In additon to the PCA plots colored by sporeformation

## Reading in the data
The data is read from the R data file create from 01_ReadInData.Rmd

```{r}
D <- readRDS("./data/pred_bacdive_growth_ribdif.rds")
D %>% select(last_col(41):last_col()) 
```

## Large analysis
### Data transformations

```{r}
# for the column pathogen, set nan to 0
Dt_tmp1 <- D %>% 
  mutate(pathogen_any = ifelse(is.na(pathogen_any), "0", pathogen_any))

# Set nan to neutrophile for PH.range
Dt_tmp2 <- Dt_tmp1 %>% 
  mutate(PH.range = ifelse(is.na(PH.range), "neutrophile", PH.range))

# Get pseudogenes as a percent
Dt_tmp3 <- Dt_tmp2 %>% 
  mutate(pseudogenes_percent = 100*(pseudogenes/total_genes))

# Calculate surface to colume ratio
Dt_tmp3.1 <- Dt_tmp3 %>% 
  mutate(motility = ifelse(is.na(motility), "Not Known", motility)) %>% 
  mutate(surface = 2*(cell.width^2) + 4*(cell.width*cell.length),
         area = cell.width*cell.width*cell.length,
         surf_area = surface/area) %>% 
          select(-surface, -area)


# Set sample as the factor seen most often for each entry
Dt_tmp3.2 <- Dt_tmp3.1 %>% 
  rowwise() %>% 
  mutate(
    max = max(pick(ends_with("counts")))
    ) %>% 
  ungroup() %>% 
  mutate(environment = case_when(
      aquatic.counts == max ~ "aquatic.counts",
      soil.counts == max ~ "soil.counts",
      plant.counts == max ~ "plant.counts",
      animal.counts == max ~ "animal.counts")) %>% 
  select(-max)

# Add surface/area for missing values based on median per family
Dt_tmp3.3 <- Dt_tmp3.2 %>% 
  # Calculate the median
  group_by(family) %>% 
  mutate(median_surf_area = median(surf_area, na.rm=T)) %>% 
  ungroup() %>% 
  # Add the median for the missng values
  mutate(surf_area = ifelse(is.na(surf_area), median_surf_area, surf_area))

# Format oxygen tolerance 
Dt_tmp3.4 <- Dt_tmp3.3 %>% 
  mutate(oxygen.tolerance = case_when(
    oxygen.tolerance == "obligate anaerobe" ~ "anaerobe",
    oxygen.tolerance == "obligate aerobe" ~ "aerobe",
    oxygen.tolerance == "facultative anaerobe" ~ "aerobe",
    oxygen.tolerance == "facultative aerobe" ~ "aerobe",
    oxygen.tolerance == "microaerophile" ~ "microaerophile",
    .default = oxygen.tolerance
  ))

# Select the data relevant for the first analysis
Dt_tmp4 <- Dt_tmp3.4 %>% 
  select(pathogen_any, PH.range, oxygen.tolerance, 
         growth_temp,  gc_percent, sporeforming,
         ar_count, surf_area, motility, environment,
         n16, phylum, ar_count, genome_components,
         pseudogenes_percent, genes_coding, species, phylum, genus,div) %>% 
  mutate(growth_temp = as.double(growth_temp)) 

# Removing the missing data which were removed for modelling
Dt_tmp5 <- Dt_tmp4 %>% 
  filter(if_all(everything(), ~!is.na(.x)))

# Make a dataframe containing the same data points used for modelling but with all attributes
cols <- Dt_tmp5 %>% colnames()
Dall <- Dt_tmp3.4%>% 
  filter(if_all(all_of(cols), ~!is.na(.x))) %>% 
  mutate(growth_temp = as.double(growth_temp)) %>% 
  mutate(n=n(), .by=phylum) %>% 
  filter(n > 10) %>% 
  select(-n) %>%   
  filter(!row_number() %in% c(681,305,865)) %>% 
  mutate(sporeforming = as.factor(sporeforming))


# Removing phylums with < 10 entries
Dt_tmp6 <- Dt_tmp5 %>% 
  mutate(n=n(), .by=phylum) %>% 
  filter(n > 10) %>% 
  select(-n)

# Remove the columns which were removed as outliers and set the spore forming attribute as a factor
Dt <- Dt_tmp6 %>% 
  filter(!row_number() %in% c(681,305,865)) %>% 
  mutate(sporeforming = as.factor(sporeforming))

```


## Effects
Defining the found model in a different way for easier use of ggeffects package. But everything is the same.
```{r}
# backtransform the transformed data
Dt_bt <- Dt %>% 
  mutate(sporeforming = as.factor(sporeforming),
         phylum = as.factor(phylum),
         motility = as.factor(motility),
         ar_count = as.factor(ar_count),
         pathogen_any = factor(pathogen_any)
         ) %>% 
  rename(`GC %` = gc_percent, `Genome components` = genome_components)

# reformat names for better plotting and with prettier names
Dt_bt <- Dt_bt %>% 
  mutate(environment = case_when(
    environment == "aquatic.counts" ~ "Aquatic", 
    environment == "animal.counts" ~ "Animal",
    environment == "plant.counts" ~ "Plant",
    environment == "soil.counts" ~ "Soil"), environment = factor(environment))

Dt_bt <- Dt_bt %>% 
  mutate(sporeforming = case_when(
    sporeforming == "0" ~ factor("No"), 
    sporeforming == "1" ~ factor("Yes")))

Dt_bt <- Dt_bt %>% 
  rename(
    `Optimal growth temperature` = growth_temp,
    `Antibiotic resistant` = ar_count,
    `Surface to volume ratio` = surf_area,
    `Coding genes` = genes_coding,
    Sporeforming=sporeforming,
    Motility=motility,
    Environment =environment,
    Phylum = phylum,
    `16S rDNA copy number` =n16
  )

# Fitting the model
fit <- lm(formula = log2(`16S rDNA copy number`) ~ `Optimal growth temperature` + `GC %` + Sporeforming + 
    `Antibiotic resistant` + `Surface to volume ratio` + Motility + Environment + Phylum + 
    log2(`Genome components`) + log2(`Coding genes`) + `Antibiotic resistant`:Phylum + `GC %`:log2(`Coding genes`) + 
    `GC %`:Sporeforming + log2(`Genome components`):log2(`Coding genes`) + 
    `Surface to volume ratio`:log2(`Coding genes`) + Motility:Environment, data = Dt_bt)

# Getting various statistics
Anova(fit)
summary(fit)
AIC(fit)
```

## Making various plots
### Plotting the distributions of the number of 16S gene copies and their entropy
```{r}
display.brewer.all()
p1 <- Dall %>% 
  filter(n16 < 20) %>% 
  group_by(phylum) %>% 
  mutate(n_phylum = n()) %>% 
  ungroup() %>% 
  mutate(Phylum=ifelse(n_phylum > 10, as.character(phylum), "Other (<10 entries)")) %>% 
  ggplot(aes(x=n16, fill=Phylum)) + 
    geom_histogram(bins = 20) +
    theme(legend.position = "none") +
    labs(x="16S rDNA copy number", y="Number of species") +
  scale_fill_brewer(palette="Paired")

p2 <- Dall %>% 
  filter(n16 < 20) %>% 
  group_by(phylum) %>% 
  mutate(n_phylum = n()) %>% 
  ungroup() %>% 
  mutate(Phylum=ifelse(n_phylum > 10, as.character(phylum), "Other (<10 entries)")) %>% 
  ggplot(aes(x=div, fill=Phylum)) + 
    geom_histogram(bins = 20) +
    labs(x="Intragenomic 16S rDNA entropy", y="Number of species") +
  
    scale_x_continuous(trans="log1p", breaks=c(0,10,100,400))+
  scale_fill_brewer(palette="Paired")

plot <- (p2 / p1)
plot + plot_annotation(tag_levels = "A")


```

# Plotting the above mentioned distributions with the summary statistics tables
```{r}
# Format summary table
tbl <- Dt_bt %>% 
  select(-species, -genus,-Phylum,-`16S rDNA copy number`,-div) %>% 
  rename(`Surface to volume ratio (1/Âµm)`=`Surface to volume ratio`,
         `pH range`=PH.range,
         `Oxygen tolerance` = oxygen.tolerance,
         `pseudogene percent`=pseudogenes_percent) %>% 
  mutate(pathogen_any = ifelse(pathogen_any == "1", "Yes","No")) %>% 
  rename(`Pathogen` = pathogen_any) %>% 
  mutate(`Antibiotic resistant` = ifelse(`Antibiotic resistant` == "1", "Yes","No")) %>% 
  mutate(Motility = case_when(
    Motility == "no" ~ "No",
    Motility == "yes" ~ "Yes",
    Motility == "Not Known" ~ "Not Known"
  )) %>% 
  tbl_summary() %>% 
  as_flex_table()
tbl <- flextable::as_raster(tbl)
tbl <- ggplot() + 
  theme_void() + 
  annotation_custom(grid::rasterGrob(tbl), xmin=-Inf, xmax=Inf, ymin=-Inf, ymax=Inf)
tbl + (p1/p2)+ plot_annotation(tag_levels = "A")


```



### Information about dataset
```{r}
# Finding the largest 5 orders
Dall %>% 
  group_by(order) %>% 
  summarise(n=n()) %>% 
  arrange(desc(n)) 

# Finding their % of the total entries
(140+126+85+71+66)/1193

# Finding the median 16S number + entropy joined at orders
Dall %>% 
  summarise(`16S rDNA copynumber`=mean(n16),.by=order) %>% 
  summarise(median(`16S rDNA copynumber`),quantile(`16S rDNA copynumber`, 0.25),quantile(`16S rDNA copynumber`, 0.75))

Dall %>% 
  summarise(div = mean(div),.by=order) %>% 
  summarise(median(div),quantile(div, 0.25),quantile(div, 0.75))

# Finding the median 16S number + entropy for the entire dataset
Dall %>% 
  rename(`16S rDNA copynumber`=n16) %>% 
  summarise(median(`16S rDNA copynumber`),quantile(`16S rDNA copynumber`, 0.25),quantile(`16S rDNA copynumber`, 0.75))
Dall %>% 
  summarise(median(div),quantile(div, 0.25),quantile(div, 0.75))


```
### Variance explained
```{r}
# Get R^2
summ(fit)

# remove phylum and its interactions and get R^2 
update(fit,.~. - Phylum -`Antibiotic resistant`:Phylum  ) %>% 
  summ()

# Print model table
tbl <- tbl_regression(fit) 
tbl %>%as_gt()
summary(fit)
Anova(fit) 
summ(fit)
```

## Antibiotic resistance plots AND estimated marginal means AND hypothesis test
```{r}
library(emmeans)
## Calculating the marginal means
ggemmeans(fit, back.transform = T,terms=c("Antibiotic resistant"))

## Plotting the Partial residuals + marginal means
dfpredict <- ggemmeans(fit, back.transform = T,terms=c("Phylum","Antibiotic resistant"))
p1 <- plot(dfpredict, dot.size = 0.3, residuals =   T, ci=T) +
  geom_point(aes(y=predicted, fill=group), position = position_dodge(0.25))+
  ggtitle("")

## Plotting the violin plots of the distribution
p2 <- Dt_bt %>% 
  ggplot(aes(col=`Antibiotic resistant`,y=`16S rDNA copy number`, x=Phylum)) +
  geom_violin(draw_quantiles = c(0.5)) 

p1 + p2 + plot_annotation(tag_levels = "A")


## Testing the difference in AR between phylums, through the estimated marginal means
emm <- emmeans(fit, specs = c("Antibiotic resistant","Phylum"))
pair <- pairs(emm, simple="each", type="response")
pair$`simple contrasts for Antibiotic resistant` %>% tidy()


```



## Environment and motility plots + hypothesis testing + marginal means
```{r}
## Calculating the marginal means
ggemmeans(fit, back.transform = T,terms=c("Environment"))

# Plot the estimated marginal means
dfpredict <- ggemmeans(fit, back.transform = T,terms=c("Environment","Motility"))
                       
p1 <- plot(dfpredict, dot.size = 0.3, residuals =   T, ci=T) +
  geom_point(aes(y=predicted, fill=group), position = position_dodge(0.25)) + ggtitle("")

p2 <- Dt_bt %>% 
  ggplot(aes(x=Environment,y=`16S rDNA copy number`, col=Motility)) +
  geom_violin(draw_quantiles = c(0.5))


# Marginal means and hypothesis testing
emm <- emmeans(fit, specs=c("Environment", "Motility"), type="response")
emm
pair <- pairs(emm, simple="each", type="response")
pair$`simple contrasts for Motility` %>% tidy()
p3 <- pwpp(emm, type="response", by="Environment") 

(p1 + p2) / p3 + plot_annotation(tag_levels = "A")

```

## Surface to volume plots + marginal effects  
```{r}
# Plot the predicted lines
cg <- c(2500,3500,4500)
ylab =  "Partial residuals (#16S)"
dfpredict <- ggpredict(fit, back.transform = T,terms=c("Surface to volume ratio [n=100]","Coding genes [cg]")) 
                       
p1 <- plot(dfpredict, dot.size = 0.3, residuals = T)+ 
  ggtitle("") 

# Plot the raw data
p2 <- Dt_bt %>% 
  ggplot(aes(x=`Surface to volume ratio`, y=`16S rDNA copy number`, col=`Coding genes`)) +
  geom_point() +
    scale_color_distiller(palette = "Blues") 
  
# Plot the distribution
p3 <- Dt_bt %>% 
  ggplot(aes(x=`Surface to volume ratio`, col=`Coding genes`)) +
  geom_histogram()


# Define the model with the transformations directly in the data
# To make the emtrends method work correctly
dt <- Dt_bt %>% mutate(`Coding genes` = log2(`Coding genes`),
                       `Genome components` = log2(`Genome components`),
                       )

fit_nt <- lm(formula = log2(`16S rDNA copy number`) ~ `Optimal growth temperature` + `GC %` + Sporeforming + 
    `Antibiotic resistant` + `Surface to volume ratio` + Motility + Environment + Phylum + 
    `Genome components` + `Coding genes` + `Antibiotic resistant`:Phylum + `GC %`:`Coding genes` + 
    `GC %`:Sporeforming + `Genome components`:`Coding genes` + 
    `Surface to volume ratio`:`Coding genes` + Motility:Environment, data = dt)


# Find the trend (Marginal effect)
trend <- emtrends(fit_nt,  ~ `Coding genes`, var="Surface to volume ratio", at=list(`Coding genes` = log2(seq(100,11000,100)), "Motility" = c("no")),"Antibiotic resistant" = c("0"),"Sporeforming"="No","Phylum"=c("Actinomycetota")) %>% 
  as.data.frame() %>% 
  tidy()

library(ggside)

# Plot the trend
p4 <- trend %>% 
  ggplot(aes(x=2^`Coding genes`, y= 2^`Surface to volume ratio.trend`*100-100)) +
  geom_line() +
  geom_ribbon(aes(ymin = 2^conf.low*100-100, ymax = 2^conf.high*100-100),alpha=0.5) +
  labs(x="Coding genes", y="Density |  % change per 1 increase in Surface to volume ratio") +
  scale_x_continuous(n.breaks = 10)

p4 <- p4 + geom_xsidefreqpoly(data=dt, aes(x=2^`Coding genes`),show.legend = F) +
  theme(ggside.panel.scale = .2)  +
  ggside(x.pos="bottom")

# Plot all plots together
wrap_plots(p1,p2,p4,p3) + plot_annotation(tag_levels = "A")

# Calculate the specific trends
trend %>% 
  mutate(`Coding genes` = 2^`Coding genes`)

```


## Temperature

```{r}

#Plot the partial residuals
ylab =  "Partial residuals (#16S)"
dfpredict <- ggpredict(fit, back.transform = T,terms=c("Optimal growth temperature [n=100]")) 
                       
p1 <- plot(dfpredict, dot.size = 1, residuals = T)+ 
  ggtitle("") + 
  xlab("Optimal growth temperature (Â°C)")

# Plot the raw data
p2 <- Dt_bt %>% 
  ggplot(aes(y=`16S rDNA copy number`, x=`Optimal growth temperature`)) +
  geom_hex() +
  xlab("Optimal growth temperature (Â°C)")
  

# Plot the raw data distributions
p3 <- Dt_bt %>% 
  ggplot(aes(x=`Optimal growth temperature`, col=`Coding genes`)) +
  geom_histogram() + 
  xlab("Optimal growth temperature (Â°C)")

p1 + (p2/p3) + plot_annotation(tag_levels = "A")


# Getting the marginal effect
trend <- emtrends(fit_nt, ~`Optimal growth temperature`,
                  var="Optimal growth temperature",
                  "Motility" = c("no"),
                  "Antibiotic resistant" = c("0"),
                  "Sporeforming"="No",
                  "Phylum"=c("Actinomycetota"))
trend %>% as.data.frame() %>% tidy()

# And plotting temperature for the ones which are sporeforming
Dt_bt %>% 
  ggplot(aes(y=`16S rDNA copy number`, x=`Optimal growth temperature`, col=Sporeforming)) +
  geom_point() +
  scale_color_discrete()  + 
  xlab("Optimal growth temperature (Â°C)")
```


## Sporeformation
```{r}
p1 <- Dt_bt %>% 
  filter(Phylum %in% c("Actinomycetota", "Bacillota")) %>% 
  ggplot(aes(col=Sporeforming,y=`16S rDNA copy number`, x=Phylum)) +
  geom_violin(draw_quantiles = c(0.5))

dfpredict <- ggpredict(fit, back.transform = T,terms=c("GC % [n=100]", "Sporeforming")) 
p2 <- plot(dfpredict, dot.size = 0.3, residuals = T)+ 
  ggtitle("") 

p3 <- Dt_bt %>% 
  ggplot(aes(x=`GC %`, y=`16S rDNA copy number`, col=`Sporeforming`)) +
  geom_point() 

p2 + (p3/p1) + plot_annotation(tag_levels ="A")


# Marginal means
emmean <- emmeans(fit_nt, pairwise ~ Sporeforming|`GC %`, var="GC %",at=list("GC %" = c(30,40,50,60)))
emmean
emmean$contrasts %>% tidy()
emmean$emmeans %>% 
  as.data.frame %>% 
  tidy() %>% 
  mutate(across(c(estimate,conf.low, conf.high), ~2^.x))

# Marginal effects and contrasts betweene them
emmipf,effort~hours,at=mylist, CIs=TRUE, plotit=FALSE)


em <- emtrends(fit_nt, pairwise ~ Sporeforming, var="GC %","GC %" = c(40,50,60))
em$contrasts %>% tidy()
em$emtrends
emmip(fit, Sporeforming~`GC %`, cov.reduce=range, CIs=T) 
em$emtrends %>% 
  as.data.frame %>% 
  tidy() %>% 
  mutate(across(c(`GC %.trend`,conf.low, conf.high), ~2^.x*100-100))
```




## Coding genes plots and estimated marginal effects
```{r}
# Plot the predicted lines, this time by building the method ourselves.

# Firstly define the data transformed and fit the model again
library(marginaleffects)
dt <- Dt_bt %>% mutate(`Coding genes` = log2(`Coding genes`),
                       `Genome components` = log2(`Genome components`),
                       )

fit <- lm(formula = log2(`16S rDNA copy number`) ~ `Optimal growth temperature` + `GC %` + Sporeforming + 
    `Antibiotic resistant` + `Surface to volume ratio` + Motility + Environment + Phylum + 
    `Genome components` + `Coding genes` + `Antibiotic resistant`:Phylum + `GC %`:`Coding genes` + 
    `GC %`:Sporeforming + `Genome components`:`Coding genes` + 
    `Surface to volume ratio`:`Coding genes` + Motility:Environment, data = dt)

# Then make a dataframe for predictions
df <- datagrid("Coding genes" = log2(seq(500, 10000, 100)) ,"GC %" = c(40,50,60),Sporeforming="No","Antibiotic resistant"= "0","Motility"="no","Phylum"="Actinomycetota",model = fit)


# Make predictions based on the dataframe
pred <- predict(fit, newdata=df, interval="confidence") %>% 
  as.data.frame() %>% 
  bind_cols(df) %>% 
  rename(estimate=fit)

# Format the predictions for plotting and calculate the residuals
pred_all <- add_predictions(dt, fit) %>% 
  bind_cols(`16S rDNA copy number`=dt$`16S rDNA copy number`) %>% 
  rename(`16S rDNA copy number`=`16S rDNA copy number...20`) %>% 
  mutate(`16S rDNA copy number` = log2(`16S rDNA copy number`)) %>% 
  mutate(res = `16S rDNA copy number` - pred) 


# Find the residual values closed to the predicted values
pred <- pred %>% 
  mutate(`Coding genes` = round(2^`Coding genes`,-2)) %>% 
  rename(gc_predict=`GC %`) %>% 
  select(-`16S rDNA copy number`)
col <- pred_all %>% 
  mutate(`Coding genes` = round(2^`Coding genes`,-2)) %>% 
  left_join(pred, by="Coding genes",multiple="all") 

# Plot everything together backtransformed
p1 <- col %>% 
  select(`GC %`,gc_predict, upr,lwr,`Coding genes`,estimate,res,`16S rDNA copy number`) %>% 
  filter(round(gc_predict,1) == round(`GC %`,-1)) %>% 
  mutate(gc_predict = factor(gc_predict)) %>% 
  ggplot(aes(y=2^estimate, x=`Coding genes`,group=`gc_predict`, col=`gc_predict`)) +
  geom_ribbon(aes(ymax=2^upr, ymin=2^lwr),alpha=0.1) +
  geom_jitter(aes(y=2^estimate+res,col=`gc_predict`),alpha=0.5) +
  geom_line() +
  ylab("16S rDNA copy number")+
  labs(color="GC %")


## Plot everything with x and y scaled
# Firstly make the labels with the estimate marginal effects
label <- col %>% 
  mutate(gc_predict = factor(gc_predict)) %>% 
  group_by(gc_predict) %>% 
  summarise(`Coding genes` = max(`Coding genes`),
            estimate=max(estimate),
            label="TEST")
label$label <- c("EME=0.753","EME=0.539","EME=0.325",NA)

#  then plot them
p2 <- col %>% 
  select(`GC %`,gc_predict, upr,lwr,`Coding genes`,estimate,res,`16S rDNA copy number`) %>% 
  filter(round(gc_predict,1) == round(`GC %`,-1)) %>% 
  mutate(gc_predict = factor(gc_predict)) %>% 
  ggplot(aes(y=2^estimate, x=`Coding genes`,group=`gc_predict`, col=`gc_predict`)) +
  geom_ribbon(aes(ymax=2^upr, ymin=2^lwr),alpha=0.1) +
  geom_jitter(aes(y=2^estimate+res,col=`gc_predict`),alpha=0.5) +
  geom_line()  +
  scale_x_continuous(trans="log2") +
  scale_y_continuous(trans="log2") +
  geom_label(aes(label=label),data=label) +
  ylab("16S rDNA copy number")+
  labs(color="GC %")

# Plot the raw data
p3 <- Dt_bt %>% 
  ggplot(aes(y=`16S rDNA copy number`, x=`Coding genes`, col=`GC %`)) +
  geom_point() +
  facet_wrap(~Phylum) +
  scale_color_distiller(palette = "Blues")
  
p3 +((p1/p2)) +plot_annotation(tag_levels = "A")


# Find the Marginal effects again wit CI
emtrends(fit,  ~ `GC %`, var="Coding genes", at=list(`GC %` = c(40,50,60), "Motility" = c("no")),"Antibiotic resistant" = c("0"),"Sporeforming"="No","Phylum"=c("Actinomycetota"))

```


## Genome components plots

```{r}
library(gghighlight)
 
## Like for coding genes, make prediction lines based on the model
# Calculate the datagrid used for predictions
df <- datagrid("Coding genes" = log2(c(2500,3500,4500)) ,"GC %" = c(50),Sporeforming="No","Antibiotic resistant"= "0","Motility"="no","Phylum"="Actinomycetota","Genome components" =log2(seq(1,16,1)),model = fit)
# Make the predictions based on the selected datagrid
pred <- predict(fit, newdata=df, interval="confidence") %>% 
  as.data.frame() %>% 
  bind_cols(df) %>% 
  rename(estimate=fit)
# Make predictions for the raw data
pred_all <- add_predictions(dt, fit) %>% 
  bind_cols(`16S rDNA copy number`=dt$`16S rDNA copy number`) %>% 
  rename(`16S rDNA copy number`=`16S rDNA copy number...20`) %>% 
  mutate(`16S rDNA copy number` = log2(`16S rDNA copy number`)) %>% 
  mutate(res = `16S rDNA copy number` - pred) 

# Format the predictions for combining them
pred <- pred %>% 
  mutate(`Genome components` = round(2^`Genome components`,0)) %>% 
  rename(cg_predict=`Coding genes`) %>% 
  select(-`16S rDNA copy number`)

# Combine the predictions from the raw data and from the grid
# Based on where genome components are the same in the datagrid and the rawdata
col <- pred_all %>% 
  mutate(`Genome components` = round(2^`Genome components`,0)) %>% 
  left_join(pred, by="Genome components",multiple="all") 

# Plot the partial residuals, and filter out the raw values which do not have values close enough to the values used for prediction.
p1 <- col %>% 
  select(cg_predict, `Coding genes`,upr,lwr,estimate,res,`16S rDNA copy number`,`Genome components`) %>% 
  filter(round(cg_predict,-2) == round(`Coding genes`,-2)) %>% 
  mutate(cg_predict = factor(2^cg_predict)) %>% 
  ggplot(aes(y=2^estimate, x=`Genome components`,group=`cg_predict`, col=`cg_predict`)) +
    geom_ribbon(aes(ymax=2^upr, ymin=2^lwr),alpha=0.1) +
    geom_jitter(aes(y=2^estimate+res,col=`cg_predict`),col="black",alpha=0.3) +
    geom_line(col="red")  +
    geom_point(col="red") +
    gghighlight(cg_predict != 1,label_key = F) +
    facet_wrap(~cg_predict, dir="v")+
  ylab("16S rDNA copy number") + labs(col="Coding genes")

#  plot the raw data
p2 <- Dt_bt %>% 
  ggplot(aes(y=`16S rDNA copy number`, x=`Genome components`, col=`Coding genes`)) +
  geom_point() +
  scale_color_distiller(palette = "Blues") + 
  scale_x_continuous(trans="log2")
  
# Plot the distribution of the raw data
p3 <- Dt_bt %>% 
  ggplot(aes(x=`Genome components`, col=`Coding genes`)) +
  geom_histogram()



p1 + (p2/p3) + plot_annotation(tag_levels = "A")


# Find the trend (Marginal effect)
emtrends(fit,  ~ `Coding genes`, var="Genome components", at=list(`Coding genes` = log2(c(1193,2500, 3500, 4500)), "Motility" = c("no")),"Antibiotic resistant" = c("0"),"Sporeforming"="No","Phylum"=c("Actinomycetota"))

# Plot the trend for a list of values for coding genes
trend <- emtrends(fit,  ~ `Coding genes`, var="Genome components", at=list(`Coding genes` = log2(seq(100,11000,100)), "Motility" = c("no")),"Antibiotic resistant" = c("0"),"Sporeforming"="No","Phylum"=c("Actinomycetota")) %>% 
  as.data.frame() %>% 
  tidy()

p3 <- trend %>% 
  ggplot(aes(x=2^`Coding genes`, y= 2^`Genome components.trend`*100-100)) +
  geom_line() +
  geom_ribbon(aes(ymin = 2^conf.low*100-100, ymax = 2^conf.high*100-100),alpha=0.5) +
  labs(x="Coding genes", y="% change per doubling of Genome components")
p1 + (p2/p3) + plot_annotation(tag_levels = "A")


```



## PCA plot
```{r}
library(ggfortify)

# Pick the attributes to plot
for_plot <- c("Sporeforming","Phylum")

# Formatting the data for plotting
D <- bind_cols(Dt_bt, select(Dall,total_seq_length)) %>%
  rename(`Sequence length` = total_seq_length)
D_subset <- D %>%
  select(`16S rDNA copy number`,`GC %`,Sporeforming, `Genome components`, `Percent pseudogenes`=pseudogenes_percent,`Antibiotic resistant`,`Optimal growth temperature`,`Sequence length`,Phylum) %>% 
  select(-`Optimal growth temperature`) %>% 
  filter(Phylum %in% c("Bacillota","Actinomycetota")) %>% 
  select(where(is.double),all_of(for_plot)) %>% 
  filter(if_all(where(is.double), ~!is.na(.x))) 

# Runnin the PCA
pca_res <- D_subset %>% 
  select(-any_of(for_plot)) %>% 
  prcomp(scale=T, center = T)

# Plotting the results
p1 <- autoplot(pca_res, loadings = TRUE, loadings.label = TRUE, 
         loadings.label.size  = 3, data = D_subset, col="Sporeforming", shape="Phylum", loadings.label.repel=T,size=2)  + scale_color_brewer(palette = "Set3")

# Plotting the raw data
p2 <- D %>% 
  filter(Phylum %in% c("Bacillota","Actinomycetota")) %>% 
  ggplot(aes(x=`Sequence length`,`16S rDNA copy number`,col=Sporeforming, shape = Phylum)) +
        geom_point() +
        facet_wrap(~Phylum,dir="v") +
        scale_color_brewer(palette = "Set3") +
        theme(legend.position = "none") +
        xlab("Sequence length (bp)")
p2+p1 + plot_annotation(tag_levels = "A")
```

## Model assumptions
```{r}
# Plot the qq plot and the residual plot
fort <- fortify(fit)
p1 <- fort %>% 
  ggplot(aes(y=.resid, x=.fitted)) +
  geom_point() +
  xlab("Fitted value") + ylab("Residual")
p2 <- fort %>% 
  ggplot(aes(sample=.stdresid)) +
  geom_qq_line() +
  stat_qq() +
  xlab("Theoretical quantile") + ylab("Standardized residual quantile")
p1+p2 + plot_annotation(tag_levels = "A")
```



