---
title: ""
output:
  pdf_document: default
  html_document: default
date: "2023-03-29"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(car)
library(tidyverse)
library(magrittr)
library(tidymodels)
library(cowplot)
library(modelr)
library(naniar)
library(ggeffects)
library(patchwork)
library(jtools)
```
## Purpose

The goal of this section is to model the intragenomic sequence diversity of the 16S gene copies. 
I addition, this model will be analysed, and the relevant attributes plotted


## Reading in the data

```{r}
D <- readRDS("./data/pred_bacdive_growth_ribdif.rds")
```

## Large analysis
### Data transformations

```{r}
# for the pathogen column, set nan to 0
Dt_tmp1 <- D %>% 
  mutate(pathogen_any = ifelse(is.na(pathogen_any), "0", pathogen_any))

# Set nan to neutrophile for PH.range
Dt_tmp2 <- Dt_tmp1 %>% 
  mutate(PH.range = ifelse(is.na(PH.range), "neutrophile", PH.range))

# Set pseudogenes as a percent
Dt_tmp3 <- Dt_tmp2 %>% 
  mutate(pseudogenes_percent = 100*(pseudogenes/total_genes))

# Calculate the surface to vollume ratio, set motility data which is missing to "Not Known"
Dt_tmp3.1 <- Dt_tmp3 %>% 
  mutate(motility = ifelse(is.na(motility), "Not Known", motility)) %>% 
  mutate(surface = 2*(cell.width^2) + 4*(cell.width*cell.length),
         area = cell.width*cell.width*cell.length,
         surf_area = surface/area) %>% 
          select(-surface, -area)


# Group sample information into the environment which each entry is sampled from predominantly
Dt_tmp3.2 <- Dt_tmp3.1 %>% 
  rowwise() %>% 
  mutate(
    max = max(pick(ends_with("counts")))
    ) %>% 
  ungroup() %>% 
  mutate(environment = case_when(
      aquatic.counts == max ~ "aquatic.counts",
      soil.counts == max ~ "soil.counts",
      plant.counts == max ~ "plant.counts",
      animal.counts == max ~ "animal.counts")) %>% 
  select(-max)
# although this will bias the element towards what is sampled more often eg. animals :D


# Add surface/volume for missing values based on median per family
Dt_tmp3.3 <- Dt_tmp3.2 %>% 
  # Calculate the median
  group_by(family) %>% 
  mutate(median_surf_area = median(surf_area, na.rm=T)) %>% 
  ungroup() %>% 
  # Add the median for the missng values
  mutate(surf_area = ifelse(is.na(surf_area), median_surf_area, surf_area))

# Format oxygen tolerance
Dt_tmp3.4 <- Dt_tmp3.3 %>% 
  mutate(oxygen.tolerance = case_when(
    oxygen.tolerance == "obligate anaerobe" ~ "anaerobe",
    oxygen.tolerance == "obligate aerobe" ~ "aerobe",
    oxygen.tolerance == "facultative anaerobe" ~ "aerobe",
    oxygen.tolerance == "facultative aerobe" ~ "aerobe",
    oxygen.tolerance == "microaerophile" ~ "microaerophile",
    .default = oxygen.tolerance
  ))

# Select the data relevant for the first analysis
Dt_tmp4 <- Dt_tmp3.4 %>% 
  select(pathogen_any, PH.range, oxygen.tolerance, 
         growth_temp,  gc_percent, sporeforming,
         ar_count, surf_area, motility, environment,
         n16, phylum, ar_count, genome_components,
         pseudogenes_percent, genes_coding,div) %>% 
  mutate(growth_temp = as.double(growth_temp)) 


```

```{r}

# Removing the missing data for modelling
Dt_tmp5 <- Dt_tmp4 %>% 
  filter(if_all(everything(), ~!is.na(.x)))

# Removing phyla with < 10 entries
Dt <- Dt_tmp5 %>% 
  mutate(n=n(), .by=phylum) %>% 
  filter(n > 10) %>% 
  select(-n)


```



```{r}
# Transforming variables
Dmodel <- Dt  %>% 
  mutate(across(c(genome_components, pseudogenes_percent, genes_coding), log2)) %>% mutate(sporeforming = factor(sporeforming)) %>% 
  mutate(div = log2(div+0.1), n16 = log2(n16))

# Remove conding genes from the model, as it is aliased with the number of 16S genes
Dmodel2 <- Dmodel %>% select(-genes_coding) %>% 
  mutate(pathogen_any = factor(pathogen_any),
         ar_count = factor(ar_count))

# Fit and select the model
fit <- lm(div ~ . , data=Dmodel2)
fit2 <- stats::step(fit, scope=~.^4, direction = "both", test="F", k=log(nrow(Dt)))

# Remove outliers
Dt_no_outliers <- Dmodel2 %>% filter(!row_number() %in% c(294))
fit <- lm(div ~ . , data=Dt_no_outliers)
fit2 <- stats::step(fit, scope=~.^4, direction = "both", test="F", k=log(nrow(Dt)))
summary(fit2)

# Run ANCOVA on the model
plot(fit2)
Anova(fit2)

# Model summary
summary(fit2)
summ(fit2)

# VIP plot
vip(fit2)
```

Fit the same model, differently with prettier names for plotting
```{r}
# Format the data correctly with pretty names
Dpretty <- Dt %>% mutate(across(where(is.character), factor)) %>% 
  filter(!row_number() %in% c(294)) %>% 
  rename(
    `Optimal growth temperature` = growth_temp,
    `Antibiotic resistant` = ar_count,
    `16S rDNA copy number` = n16,
    `Intragenomic 16S gene entropy`=div,
    `Genome components`=genome_components,
    `Surface to volume ratio` = surf_area)

# Fit the model
fitpretty <- lm(formula = log2(`Intragenomic 16S gene entropy`+0.1) ~ `Optimal growth temperature` +`Antibiotic resistant` +`Surface to volume ratio` + log2(`16S rDNA copy number`) + log2(`Genome components`) + 
    `Antibiotic resistant`:log2(`Genome components`), data = Dpretty)

# Model summary
summ(fitpretty,pvals = F,confint = T)

# Model diagnostics
par(mfrow=c(2,2))
plot(fitpretty)

# VIP plot
vip <- vip(fitpretty)


# Plot predicted lines and raw dataplots in addition to VIP plot.
sizedot <- 1
reson = F
jitter = 0
ylab01 = "Entropy+0.1"
geomtype <- geom_hex()
ylab ="Entropy"
Dprettyplot <- Dpretty %>% filter(`Intragenomic 16S gene entropy` < 50)

dfpredict <- ggpredict(fitpretty, back.transform = T, terms=c("Optimal growth temperature [all]"))
TempP <- plot(dfpredict, dot.size = sizedot, ci=T, residuals = reson, jitter=jitter)  + ggtitle("")+ ylab(label = ylab01)
TempR <- Dprettyplot %>% ggplot(aes(y=`Intragenomic 16S gene entropy`,x=`Optimal growth temperature`)) + geomtype + ylab(label = ylab)

dfpredict <- ggpredict(fitpretty, back.transform = T, terms=c("Surface to volume ratio [all]"))
surfP <- plot(dfpredict, dot.size = sizedot, ci=T, residuals = reson, jitter=jitter)  + ggtitle("")+ ylab(label = ylab01)
surfR <- Dprettyplot %>% ggplot(aes(y=`Intragenomic 16S gene entropy`,x=`Surface to volume ratio`)) + geomtype+ ylab(label = ylab)

dfpredict <- ggpredict(fitpretty, back.transform = T, terms=c("16S rDNA copy number [all]"))
s16P <- plot(dfpredict, dot.size = sizedot, ci=T, residuals = reson, jitter=jitter)  + ggtitle("")+ ylab(label = ylab01)
s16R <- Dprettyplot %>% ggplot(aes(y=`Intragenomic 16S gene entropy`,x=`16S rDNA copy number`)) + geomtype+ ylab(label = ylab)

dfpredict <- ggpredict(fitpretty, back.transform = T, terms=c("Genome components [all]", "Antibiotic resistant"))
gc_arP <- plot(dfpredict, dot.size = sizedot, ci=T, residuals = reson, jitter=jitter)  + ggtitle("")+ ylab(label = ylab01)
gc_arR <- Dprettyplot %>% ggplot(aes(y=`Intragenomic 16S gene entropy`,x=`Genome components`, col=`Antibiotic resistant`)) + geom_point(size=0.5)+ ylab(label = ylab)


wrap_plots(s16P, s16R,TempP , TempR,ncol = 2,tag_level = "new") +plot_annotation(tag_levels = "A") 
wrap_plots(surfP , surfR,gc_arP,gc_arR,ncol = 2,tag_level = "new") +plot_annotation(tag_levels = "A") 
vip

```

```{r}
# Get model summary
summ(fitpretty)

```


```{r}
# Get the estimated marginal effect of the logdiv~logn16 relationship
emtrends(fit2, var = "n16",specs = "n16") %>% 
  as.data.frame() %>%
  tidy() %>% 
  mutate(across(c(n16.trend,conf.low,conf.high),~100*2^.x-100))

# Get the estimated marginal effect of the temperatures effect
emtrends(fit2, var = "growth_temp",specs = "growth_temp") %>% 
  as.data.frame() %>%
  tidy() %>% 
  mutate(across(c(growth_temp.trend,conf.low,conf.high),~100*2^.x-100))

# and for surf area effect
emtrends(fit2, var = "surf_area",specs = "surf_area") %>% 
  as.data.frame() %>%
  tidy() %>% 
  mutate(across(c(surf_area.trend,conf.low,conf.high),~100*2^.x-100))

# And for the arcount~genome_components term
emtrends(fit2, var = "genome_components",specs = "ar_count") %>% 
  as.data.frame() %>%
  tidy() %>% 
  mutate(across(c(genome_components.trend,conf.low,conf.high),~100*2^.x-100))
```


```{r}
# Lastly get the contrast between non antibiotic resistant bacteria and antibiotic resistant bacteria
em <- emmeans(fit2, pairwise ~ ar_count | genome_components,at=list("genome_components" = at))
em$contrasts %>% 
  confint() 
em$contrasts


```






















